### PR TITLE
Introduce --dump-build-time flag to BuildCommand

### LINKED ISSUE #31238: Create a dashboard to monitor build time
**Description:**
A dashboard should be created to monitor the changes in build time using a few samples like hello-world, a connector etc.

### PR DESCRIPTION
**Pull request by azinneera**
Fixes #31238 

## Approach
`bal build --dump-build-time` will generate `target/built-time.json` with content simlar to following:
```json
{
    "timestamp": 1624304381716,
    "offline": false,
    "projectLoadDuration": 435,
    "packageResolutionDuration": 2579,
    "packageCompilationDuration": 175,
    "codeGenDuration": 547,
    "testingExecutionDuration": 0,
    "totalDuration": 3763
}
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by suganyasuven** [comment]
Shall we update this `distribution/zip/jballerina-tools/resources/command-completion/command-completion.csv` as well?

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java [hunk 1]
```diff
@@ -69,6 +71,15 @@ public BuildCommand() {
         this.skipCopyLibsFromDist = false;
     }
 
+    public BuildCommand(Path projectPath, PrintStream outStream, PrintStream errStream, boolean dumpBuildTime) {
+        this.projectPath = projectPath;
+        this.outStream = outStream;
+        this.errStream = errStream;
+        this.exitWhenFinish = false;
+        this.skipCopyLibsFromDist = true;
```
Comment (pramodya1994) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java:79]: Can we remove this?

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/task/DumpBuildTimeTask.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                throw createLauncherException("couldn't write build time to file : " + e.getMessage());
+            }
+        } catch (IOException e) {
+            throw createLauncherException("couldn't write build time to file : " + e.getMessage());
+        }
+    }
+
+    private Path getBuildTimeFilePath(Project project) {
+        if (project.kind().equals(ProjectKind.BUILD_PROJECT)) {
+            return project.sourceRoot().resolve("target").resolve("build-time.json").toAbsolutePath();
```
Comment (pramodya1994) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/task/DumpBuildTimeTask.java:79]: Do we need to add `build-time.json` as a constant?

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/BuildTime.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    private static final BuildTime instance = new BuildTime();
+
+    public long timestamp;
+    public boolean offline;
+    public long projectLoadDuration;
+    public long packageResolutionDuration;
+    public long packageCompilationDuration;
+    public long codeGenDuration;
+    public long testingExecutionDuration;
+    public long totalDuration;
```
Comment (pramodya1994) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/BuildTime.java:36]: Why we are not using getters and setters here? So we can remove `URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD` exclude.

**Comment by hevayo** [comment]
Can we have executable jar creation time. I mean assembling the executable jar

**Comment by azinneera** [comment]
> Can we have executable jar creation time. I mean assembling the executable jar

@hevayo Will the following format work?
```json
{
    "timestamp": 1624304381716,
    "offline": false,
    "compile": false, --> new addition to check if it is a bala/executable
    "projectLoadDuration": 435,
    "packageResolutionDuration": 2579,
    "packageCompilationDuration": 175,
    "codeGenDuration": 347,
    "emitArtifactDuration": 200, --> bala/exec creation time
    "testingExecutionDuration": 0,
    "totalDuration": 3763
}
```

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/BuildTime.java (hunk 3 above)
Comment (azinneera) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/BuildTime.java]: This class is internal to the CLI module. We don't export this in the module. So I think there is no harm in having them as public.

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/task/DumpBuildTimeTask.java (hunk 2 above)
Comment (azinneera) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/task/DumpBuildTimeTask.java]: Addressed in 995c32b

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java (hunk 1 above)
Comment (azinneera) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java]: Addressed in 995c32b

**Review by suganyasuven — APPROVED** [review]

**Review by pramodya1994 — APPROVED** [review]
LGTM

**Pull request merged**